Generalise the field type used for DataType.id from int

mz2 opened the pull request:
In this PR I have tried to address the issue noted in https://github.com/vapor/postgresql-driver/issues/17, of adding support for non-numerical primary IDs. As discussed in the thread, the problem is not really with PostgreSQL, but a limitation of the SQL serializer, and indeed on a closer look also the schema creator needed adaptation.

## Summary of changes proposed here

- Add a `KeyType` enum next to `DataType` in `Schema.Field` to allow describing the type used as an identifier. 
- Associate a `KeyType?` with the `id` case of `DataType` to optionally allow describing the kind of identifier that is to be used.
- Move the creation of a pivot field to the schema creator, as it belongs in that domain (later on one might want to add for example foreign key constraints into the schema as well?).
- Add a `KeyType?` argument where necessary to pass context to schema creation etc. 

## Impact on existing code

No existing code (that implicitly assumes a single primary ID type exists per driver) should break as a result of these changes, as all methods which now accept a `KeyType` or `KeyType?` have no new required arguments, and the existing switch/cases which deal with the id case simply continue to work (on the contrary, there may be additional changes still required for non-integral keys that I may not have foreseen). The key type given for an id is in other words just a hint that the drivers may choose to interpret in some way, or choose not to interpret (I suppose given the associated value is marked optional, it would make sense for drivers to error out if you try using a key type that is not supported, for example a numerical ID with Mongo).

## Alternatives considered

- A stringly typed associated value to `id`. This would not be good for backend agnosticism in mind. Also, because the `KeyType` which I preferred in this solution includes a `custom` case, unforeseen key types are still possible include.
- Avoid introducing a new `KeyType` by recursively associating to `DataType`. I decided against this because not all field types make sense as a primary identifier and the implementation burden to concrete providers / drivers is increased, especially if the API is to remain enough to be used with no-SQL databases as well.

## Limitations

Only string and "custom" cases are present there in `KeyType`. I figured this would be a limitation lifted in subsequent iterative changes – for example these changes are a helpful prerequisite to adding support for specifically UUID typed IDs (https://github.com/vapor/fluent/issues/107).

mz2 commented:
(If the changes are otherwise to your liking, I am happy to add the necessary code coverage of course.)

LoganWright commented on Sources/Fluent/Entity/Entity.swift:
```diff
@@ -82,6 +84,10 @@ extension Entity {
         return String(describing: self).lowercased()
     }
+    
+    public static var idType: Schema.Field.KeyType? {
```
should this be optional? Seems like we're defaulting to `.int` all over the place, maybe here could just be the point of default. Not sure of further implications somewhere else right now

tannernelson replied:
I'd prefer this to be non-optional with `.int` as a default.

mz2 replied:
Reason I proposed an optional there is this would leave it up for the driver to interpret as whatever makes sense as the default primary ID type; Integral primary IDs for example may not be possible for some drivers even. But I can see how that capability still remains available, and beside centralising the assumption it is probably also is clearer for more users of the API too.

tannernelson replied:
That's a good point. Since the entity has access to `database` could we just return `database.driver.idType` by default?

mz2 replied:
I think that makes sense.

mz2 replied:
Hmm, `database` of `Entity` is an optional, which is not ideal in making this `Entity.idType` non-optional by relying on `database?.driver.idType`. It appears other cases of dealing with that in `Entity.swift` either return nil or throw an error, neither of which would really be ideal here. When trying to resolve `idType` I would say it is reasonable to treat it as a precondition that `database != nil` (with a message telling that you've probably forgotten to set the database of the entity)?

Further, there is no `idType` on Driver yet, only `idKey`. Did you mean I should introduce `idType` in there (I am in agreement, if yes)?

LoganWright replied:
I think having this be non-optional, with something like `return database?.driver.idType ?? .int` seems like a reasonable solution

mz2 replied:
Ok, agreed. I suppose my general approach would be to fail early and decisively (preconditionFailure), but if a client of this API is attempting to for example create a schema with no database associated with the entity, they're going to find out they're doing it wrong also through some other mechanism than a precondition failure here.

LoganWright commented on Sources/Fluent/Schema/Schema+Creator.swift:
```diff
@@ -134,12 +135,34 @@ extension Schema {
+                            unique: false)
+            
+            fields += Field(name: rightName,
+                            type: rightKeyType.fieldType,
```
this shouldn't be `.id` or something here since it's a pivot? Just clarifying.

mz2 replied:
Could you clarify the question? :-) Are you asking if there's potentially a mistake here, or to adjust some local variable or property naming?

mz2 replied:
If you mean adding the `_id` suffix, I slightly reordered things around and that gets added through the "suffix" parameter this method receives. The (only) call site is above in Join.swift:

```
+            builder.pivotKeys(left: left, right: right, suffix: "_\(database.driver.idKey)")
```

LoganWright replied:
@mz2 I'm asking if there's potentially a mistake here with the fieldType since normally, it corresponds to an `idType`, but I might be misunderstanding the role within a pivot, so just wanted to clarify :)

mz2 replied:
Ah, but it's the Field initializer, whose `type:` parameter indeed is a `DataType`. So, I do think it's correct indeed.

tannernelson commented on Sources/Fluent/Schema/Schema+Creator.swift:
```diff
@@ -15,13 +15,14 @@ extension Schema {
         
         public func id(
```
There needs to be some way to only have to specify id type once.

```swift
class User: Model {
    static var idType = .uuid
    ...

    static func prepare() {
        ...
        users.id(idType: .uuid)
        ...
    }
}
```

Having to repeat `.uuid` in two places isn't ideal.

tannernelson replied:
Maybe we can add an overload like 

```swift
users.id(for: User.self)
```

mz2 replied:
Right, I think that makes sense. Could take it even further and instead of overloading, only allow passing in an entity type?

mz2 replied:
(I am suggesting a breaking change as I assumed this is for inclusion in 2.0?)

mz2 commented:
I think I've addressed all the comments. I also added some more tests to cover the newly added branches for .string and .custom typed schema creation. There may be still some bits that are not covered by tests.

mz2 commented on Sources/Fluent/Database/Driver.swift:
```diff
@@ -46,4 +54,8 @@ extension Driver {
     }
+    
+    public var idType: Schema.Field.KeyType {
+        return .int
```
I added this in an extension – on reflection, not sure if this should be there. May be better for all driver implementations to knowingly set it? given that this concept of an idType is new and some drivers may not be behaving correctly and the compiler doesn't inform you about it being necessary to declare this.

mz2 commented on Sources/Fluent/SQL/GeneralSQLSerializer.swift:
```diff
@@ -403,8 +403,17 @@ open class GeneralSQLSerializer: SQLSerializer {
+                return "INTEGER PRIMARY KEY"
+                
+            case .string(let length):
+                return "VARCHAR(\(length ?? 128)) PRIMARY KEY"
```
Pointing this out too as it may be controversial: I changed from STRING to VARCHAR because this would work across MySQL, PostgreSQL, SQLite at least whereas STRING would not I'm pretty sure.

LoganWright replied:
That seems like a good move, can you elaborate why it might be controversial?

mz2 replied:
Just because it is arguably unrelated to what this PR is about.

mz2 commented:
Ok, I've now added coverage of practically all bits that matter. There is some untested code in the test utility structs I added, but that's just because of protocol requirements.

mz2 commented:
Ready for another review.

mz2 commented:
The Linux build succeeded but macOS build failed:

```
error: exit(1): /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swift-build-tool -f /Users/travis/build/vapor/fluent/.build/debug.yaml
❌  Build failed
/Users/travis/build.sh: line 59: shell_session_update: command not found
```

wut? Is that some travis issue, or genuine?

vzsg commented:
There were more lines to that error actually (regarding final and dynamic). 😛
I restarted the build, Node 2.0.0-alpha.2 will most certainly fix it.

mz2 commented:
Yep – didn't quite get how that could contribute as it builds cleanly on local machine (so i assumed it wasn't something bad that got introduced from the merge to master) – except for a different toolchain / xcode being used? Appears happy now on 2nd attempt?

tanner0101 commented on Sources/Fluent/Schema/Schema+Field.swift:
```diff
@@ -11,14 +11,31 @@ extension Schema {
             case custom(type: String)
         }
+        
+        public enum KeyType {
```
I think we should call this `IdentifierKeyType` and make it a top level type.

mz2 replied:
Reason I proposed a type here is it'd be consistent with the `DataType` (and it is indeed a field related thing). I also wasn't sure if the type would get referenced also in other contexts beside the .id case of DataType (for example foreign key related uses, or cases where this type would be used to describe something that is not necessarily the primary key of the record, which I've understood .id refers to). No concrete issues come to mind though, happy to do the change.

LoganWright replied:
@mz2 @tanner0101 I think it's kind of confusing as just `KeyType` especially w/ `DataType` because the two look so similar. Can we call it `IdType` since that's what we're using everywhere else, or `PrimaryKeyType` to stick w/ more conventions. `case id(type: PrimaryKeyType)` reads nicely to me. (I prefer `PrimaryKey` to `Id` in general when dealing w/ types, just a thought)

mz2 replied:
Out of these options I would find `IdType` the clearest myself as it would be inclusive of other uses of it beyond primary keys (which also is a relational database term, whereas also for example Mongo is now supported). It also differentiates it from `DataType` clearly enough IMO.

tanner0101 commented:
moved to: https://github.com/vapor/fluent/pull/177

I'm going to work on including this in my PR to create better pivot functionality. Keep your eyes on: https://github.com/vapor/fluent/pull/178
